File: solutions/python/clock/1/clock.py
# Dedicated to Shree DR.MDD

DAY = 1440

class Clock:
  def __init__(self, hrs, mins):
    self.__tick = (hrs * 60 + mins) % DAY

  def __str__(self):
    return '%02d:%02d' % (self.__tick // 60, self.__tick % 60)

  def __repr__(self):
    return 'Clock(%d, %d)' % (self.__tick // 60, self.__tick % 60)

  def __eq__(self, rhs):
    return self.__tick == rhs.__tick

  def __add__(self, delta):
    return Clock(0, self.__tick + delta)

  def __sub__(self, delta):
    return Clock(0, self.__tick - delta)

File: solutions/kotlin/clock/1/src/main/kotlin/Clock.kt
// Dedicated to Shree DR.MDD
import java.time.Duration

data class Clock(private var totalMins: Int) {
    constructor(h: Int, m: Int) : this(h.times(MINS_IN_HOUR).plus(m).normalizeDayMinutes())

    fun add(extraMins: Int) {
        totalMins = totalMins.plus(extraMins).normalizeDayMinutes()
    }

    fun subtract(lessMins: Int) {
        totalMins = totalMins.minus(lessMins).normalizeDayMinutes()
    }

    override fun toString() = "${getHrs.display2Digits()}:${getMins.display2Digits()}"

    private val getHrs get() = totalMins.div(MINS_IN_HOUR)
    private val getMins get() = totalMins.safeModulo(MINS_IN_HOUR)
}

private fun Int.safeModulo(other: Int) = (this % other + other) % other
private fun Int.display2Digits() = "%02d".format(this)
private fun Int.normalizeDayMinutes() = safeModulo(MINS_IN_DAY)

private val MINS_IN_DAY = Duration.ofDays(1).toMinutes().toInt()
private val MINS_IN_HOUR = Duration.ofHours(1).toMinutes().toInt()
